### PR TITLE
O3-953: Location picker does not show all locations

### PR DESCRIPTION
**Pull request by vasharma05**
## Requirements
- [x] This PR has a title that briefly describes the work done, including the ticket number if there is a ticket.

#### For changes to apps
- [x] My work conforms to the [**OpenMRS 3.0 Styleguide**](https://om.rs/styleguide).

#### If applicable
- [ ] My work includes tests or is validated by existing tests.
- [ ] I have updated the [esm-framework mock](https://github.com/openmrs/openmrs-esm-core/blob/master/packages/framework/esm-framework/mock.tsx) to reflect any API changes I have made.

## Summary
This PR is still a work in progress.
I have added infinite scroll to fetch the multiple sets of locations in the location picker. Along with that, I have added backend connectivity to the search bar.

## Screenshots
https://www.awesomescreenshot.com/video/7704971?key=67e59ee20d21e2a45aac94fd502e19d9

## Related Issue
https://issues.openmrs.org/browse/O3-953

### CODE REVIEW

**Comment by ibacher** [comment]
@vasharma05 Very cool demo!

**Review by ibacher — COMMENTED** [review]
A few small comments, but good work so far!

File: packages/apps/esm-login-app/src/choose-location/choose-location.component.tsx [hunk 1]
```diff
@@ -54,32 +55,21 @@ export const ChooseLocation: React.FC<ChooseLocationProps> = ({
...
-      return () => ac.abort();
-    }
-  }, [isLoginEnabled]);
-
-  useEffect(() => {
-    if (loginLocations) {
-      if (!config.chooseLocation.enabled || loginLocations.length < 2) {
-        changeLocation(loginLocations[0]?.resource.id);
+    if (locationData) {
+      if (!config.chooseLocation.enabled || locationData.length < 2) {
```
[the remark this answers is not in the document]
Comment (vasharma05) [reply packages/apps/esm-login-app/src/choose-location/choose-location.component.tsx]: Sure! No worries!

File: packages/apps/esm-login-app/src/choose-location/choose-location.resource.ts [hunk 2]
```diff
@@ -1,40 +1,70 @@
...
-import { LocationResponse } from "../types";
+import { LocationEntry, LocationResponse } from "../types";
+import useSwr from "swr";
 
-export function getLoginLocations(): Observable<Object[]> {
-  return openmrsObservableFetch(
-    `/ws/rest/v1/location?tag=Login%20Location&v=custom:(uuid,display)`
-  ).pipe(map(({ data }) => data["results"]));
-}
+export const fhirLocationUrl = `${fhirBaseUrl}/Location?_summary=data`;
```
[the remark this answers is not in the document]
Comment (vasharma05) [reply packages/apps/esm-login-app/src/choose-location/choose-location.resource.ts]: Actually, I used it as an import but then removed it, forgot to remove it from here.
Thanks!

File: packages/apps/esm-login-app/src/choose-location/choose-location.resource.ts [hunk 3]
```diff
@@ -1,40 +1,70 @@
...
-  useLoginLocationTag: boolean
-) {
-  return searchLocationsFhir(
-    location,
-    abortController,
-    useLoginLocationTag
-  ).then((locs) => locs.data.entry);
+  // URL fetching
+  let url = fhirLocationUrl;
+  url = url.concat(`&_count=${count}`);
```
[the remark this answers is not in the document]
Comment (vasharma05) [reply packages/apps/esm-login-app/src/choose-location/choose-location.resource.ts]: Duly noted.
Thanks alot!

File: packages/apps/esm-login-app/src/choose-location/choose-location.resource.ts [hunk 4]
```diff
@@ -1,40 +1,70 @@
...
+    setLocationData((locationData) =>
+      swrResult.data
+        ? locationData
+          ? [...new Set([...locationData, ...swrResult.data?.data?.entry])]
+          : swrResult.data?.data?.entry
+        : locationData
+    );
+    if (swrResult.data) {
+      setTotalResults(swrResult?.data?.data?.total);
+    }
```
[the remark this answers is not in the document]
Comment (vasharma05) [reply packages/apps/esm-login-app/src/choose-location/choose-location.resource.ts]: A lot of clean-up was pending at that time.
I have done some cleaning, please do review the new work too.
Thanks for pointing this out.

**Comment by vasharma05** [comment]
I had a conversation with @denniskigen about helping me out with the writing tests.
We'll be having a look at the tests tomorrow.

**Comment by vasharma05** [comment]
Hi @ZacButko!
Can you please review the above 3 commits, I have replaced my code with SWR Infinite.
Thanks!

File: packages/apps/esm-login-app/src/choose-location/choose-location.resource.ts [hunk 5]
```diff
@@ -1,40 +1,85 @@
...
+    showNotification({
+      title: error.name,
+      description: error.message,
+      kind: "error",
+    });
+  }
+
+  useEffect(() => {
+    setSize(1);
+  }, [searchQuery]);
```
Comment (ZacButko) [inline packages/apps/esm-login-app/src/choose-location/choose-location.resource.ts:65]: Missing dependency 'setSize'

File: packages/apps/esm-login-app/src/choose-location/choose-location.resource.ts [hunk 6]
```diff
@@ -1,40 +1,85 @@
...
+      totalResults: data?.[0]?.data?.total ?? null,
+      hasMore: data?.length
+        ? data?.[data.length - 1]?.data?.link.some(
+            (link) => link.relation === "next"
+          )
+        : false,
+      loadingNewData: isValidating,
+      setPage: setSize,
+    };
+  }, [data, isValidating, setSize, size]);
```
Comment (ZacButko) [inline packages/apps/esm-login-app/src/choose-location/choose-location.resource.ts:82]: Unnecessary dependency 'size'

File: packages/apps/esm-login-app/src/location-picker/location-picker.component.tsx [hunk 7]
```diff
@@ -155,30 +97,53 @@ const LocationPicker: React.FC<LocationPickerProps> = ({
...
+            setPage((page) => page + 1);
+          }
+        },
+        {
+          threshold: 1,
+        }
+      );
+      if (node) observer.current.observe(node);
+    },
+    [loadingNewData, hasMore]
```
Comment (ZacButko) [inline packages/apps/esm-login-app/src/location-picker/location-picker.component.tsx:135]: Missing dependency setPage

File: packages/apps/esm-login-app/src/location-picker/location-picker.component.tsx [hunk 8]
```diff
@@ -155,30 +97,53 @@ const LocationPicker: React.FC<LocationPickerProps> = ({
...
-      ? setPageSize(locationData.locationResult.length)
-      : setPageSize(chooseLocation.numberToShow);
-  }, [locationData.locationResult.length]);
+  // Infinte scrolling
+  const observer = useRef(null);
+  const loadingIconRef = useCallback(
+    (node) => {
+      if (loadingNewData) return;
+      if (observer.current) observer.current.disconnect();
+      observer.current = new IntersectionObserver(
```
Comment (ZacButko) [inline packages/apps/esm-login-app/src/location-picker/location-picker.component.tsx:123]: 👍

**Review by ZacButko — DISMISSED** [review]
Looks good. Please fix missing / excessive hook dependencies before merging

File: packages/apps/esm-login-app/src/config-schema.ts [hunk 9]
```diff
@@ -36,6 +36,12 @@ export const configSchema = {
       _default: 8,
       _description: "The number of locations displayed on location picker",
     },
+    resultsInOneCycle: {
```
Comment (ibacher) [inline packages/apps/esm-login-app/src/config-schema.ts:39]: This is going to be a hard one to name properly... I'm not even certain if it is something we need to be user-configurable, but we can probably come up with a better name for this option.

```suggestion
    locationsPerRequest: {
```

File: packages/apps/esm-login-app/src/location-picker/location-picker.component.scss [hunk 10]
```diff
@@ -37,8 +37,9 @@
 .searchResults {
   @extend .bodyShort02;
   margin: 1rem 0rem;
-  height: 26.125rem;
+  // height: 26.125rem;
```
Comment (ibacher) [inline packages/apps/esm-login-app/src/location-picker/location-picker.component.scss:40]: Probably want to either uncomment this line or remove it.

**Review by ibacher — COMMENTED** [review]
Few more things. Nice work, though, Vineet!

**Review by ibacher — APPROVED** [review]

**Comment by ibacher** [comment]
Tests to be added in subsequent PR

**Review by ZacButko — APPROVED** [review]

**Comment by vasharma05** [comment]
Thank you @denniskigen for writing the tests at #361.

**Pull request merged**